Add BatchWrite method for Go client

pawanrawal opened the pull request:
Opening a new PR because I rebased master on the previous branch and it messed up the diff on Reviewable.

I have added support for BatchWrite method for the Go client.

### TODOS
1.  Change the Geo stuff to use this.
2. Move examples from main.go to within the code and delete the file.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/401)
<!-- Reviewable:end -->

manishrjain commented:
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[goclient/client/client.go, line 46 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPwElT5-_D8wOGXrge:-KZPwElT5-_D8wOGXrgf:begzqw8) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L46)):*
> ```Go
> // or perform set/delete mutations.
> func NewRequest() Req {
> 	return Req{}
> ```

You don't need this function.

---

*[goclient/client/client.go, line 62 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPwLIWNO7fP7VCn6qE:-KZPwLIWNO7fP7VCn6qF:b-a76zrk) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L62)):*
> ```Go
> 		return fmt.Errorf("Predicate can't be empty")
> 	}
> 	hasVal := nq.Value != nil && nq.Value.Val.(*graph.Value_StrVal).StrVal != ""
> ```

Would this always be string?

---

*[goclient/client/client.go, line 144 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPwrnHqzALFH4QyiW4:-KZPwrnHqzALFH4QyiW5:batddva) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L144)):*
> ```Go
> func (req *Req) reset() {
> 	req.gr = graph.Request{}
> 	req.gr.Mutation = new(graph.Mutation)
> ```

You could just clear it, instead of allocating new memory for this.

---

*[goclient/client/client.go, line 172 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPx-TuMYKwNN_rqj1H:-KZPx-TuMYKwNN_rqj1I:b-bw5t64) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L172)):*
> ```Go
> 	counter := atomic.AddUint64(&batch.mutations, 1)
> RETRY:
> 	ctx, _ := context.WithTimeout(context.Background(), time.Minute)
> ```

You should be calling the cancel to release memory.

---

*[goclient/client/client.go, line 190 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPxitKHPwVZzflQ0tX:-KZPxitKHPwVZzflQ0tY:b-o9otct) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L190)):*
> ```Go
> 			batch.request(req)
> 		}
> 	}
> ```

You want to do the batch.request in this goroutine, after it exits the for loop.

---

*[goclient/client/client.go, line 198 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPynn7rmac_NdXfLSv:-KZPynn7rmac_NdXfLSw:bv8wbzo) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L198)):*
> ```Go
> 		size:    size,
> 		pending: pending,
> 		nquads:  make(chan NQuadOp, 3*(pending)),
> ```

batchsize*pending or batchsize*2 or something.

---

*[goclient/client/client.go, line 238 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPxs9KXteGLoQy-d7W:-KZPxs9KXteGLoQy-d7X:b-j83gnu) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L238)):*
> ```Go
> 	close(batch.nquads)
> 	for _, req := range batch.requests {
> 		batch.request(req)
> ```

This shouldn't be required. You just want to wait here for all the goroutines to finish.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/401)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 7 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[goclient/client/client.go, line 46 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPwElT5-_D8wOGXrge:-KZQ8MV_HnTXPzsF4zIa:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L46)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You don't need this function.
</blockquote></details>

Done.

---

*[goclient/client/client.go, line 62 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPwLIWNO7fP7VCn6qE:-KZQ6MMAaRn_cbcCsi-j:buo93w2) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L62)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Would this always be string?
</blockquote></details>

No, I will change this when I make the change for removing rdf.NQuad. Right now graph.NQuad and rdf.NQuad aren't the same. Added a TODO.

---

*[goclient/client/client.go, line 144 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPwrnHqzALFH4QyiW4:-KZQ8OgnqiyXbV8kG4FN:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L144)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You could just clear it, instead of allocating new memory for this.
</blockquote></details>

Done.

---

*[goclient/client/client.go, line 172 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPx-TuMYKwNN_rqj1H:-KZQ8Zen-ZYyrIRZzjwo:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L172)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You should be calling the cancel to release memory.
</blockquote></details>

Done.

---

*[goclient/client/client.go, line 190 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPxitKHPwVZzflQ0tX:-KZQ8_YqA6kCgOfQqLNf:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L190)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You want to do the batch.request in this goroutine, after it exits the for loop.
</blockquote></details>

Done.

---

*[goclient/client/client.go, line 198 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPynn7rmac_NdXfLSv:-KZQ8bxd08OBKghnidQu:b5t10f5) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L198)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

batchsize*pending or batchsize*2 or something.
</blockquote></details>

Kept it at 2*size for now.

---

*[goclient/client/client.go, line 238 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZPxs9KXteGLoQy-d7W:-KZQ8emED06IUKxFZaGV:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/99e7294bd3c727b51c4e86e3d1f5ddf0fa663138/goclient/client/client.go#L238)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This shouldn't be required. You just want to wait here for all the goroutines to finish.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/401)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 13 of 16 files at r2.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[cmd/dgraphloader/main.go, line 39 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYpvwp8wd-Zdf9BM_J:-KZYpvwp8wd-Zdf9BM_K:by89iqc) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L39)):*
> ```Go
> 	files   = flag.String("r", "", "Location of rdf files to load")
> 	dgraph  = flag.String("d", "127.0.0.1:8080", "Dgraph server address")
> 	geoJSON = flag.String("json", "", "Json file from which to upload geo data")
> ```

geojson

---

*[cmd/dgraphloader/main.go, line 110 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYwPAQjOPOmowhhQYF:-KZYwPAQjOPOmowhhQYG:bnchovw) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L110)):*
> ```Go
> 
> 	var r io.Reader
> 	r = f
> ```

You don't need this.

---

*[cmd/dgraphloader/main.go, line 123 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYwbjjo-3kFRX9_f7E:-KZYwbjjo-3kFRX9_f7F:b-c9i8jq) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L123)):*
> ```Go
> // Upload a geojson file to the server by converting it into the NQuads format.
> func Upload(batch *client.BatchMutation, r io.Reader) {
> 	cfeat := make(chan result, 1000)
> ```

Move this func insided the uploadJSON.

---

*[cmd/dgraphloader/main.go, line 134 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYwi6YUPjg4SI1YOlo:-KZYwi6YUPjg4SI1YOlp:behopz0) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L134)):*
> ```Go
> 
> // ValueFromJson converts geojson into a client.Value
> func ValueFromJson(json string) (client.Value, error) {
> ```

You don't need this func.

---

*[cmd/dgraphloader/main.go, line 166 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZ-3w63mydvB22WfbQ:-KZZ-3w7D0vXe6SbD1Hm:b-8sr85b) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L166)):*
> ```Go
> }
> 
> func toMutations(f *geojson.Feature, batch *client.BatchMutation) {
> ```

Just get rid of all special geo stuff here for now.

---

*[cmd/dgraphloader/main.go, line 232 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYwxCR-ArPOZXfipdn:-KZYwxCR-ArPOZXfipdo:b-4xs1rp) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L232)):*
> ```Go
> }
> 
> func parse(r io.Reader, c chan<- result) {
> ```

parseGeoJSON

---

*[cmd/dgraphloader/main.go, line 291 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYvxcsChfOMZWFN5dn:-KZYvxcsChfOMZWFN5do:b7xzeyi) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L291)):*
> ```Go
> 	var err error
> 	conn, err := grpc.Dial(*dgraph, grpc.WithInsecure())
> 	if err != nil {
> ```

x.Checkf(..)

---

*[cmd/dgraphloader/main.go, line 296 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYvuhn2UYH_t_Hid_9:-KZYvuhn2UYH_t_Hid_A:b-m0tqv1) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L296)):*
> ```Go
> 	defer conn.Close()
> 
> 	batch := client.NewBatchMutation(context.Background(), conn, 1000, 10)
> ```

The last 2 args were part of flags.

---

*[cmd/dgraphloader/main.go, line 301 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYw2GbsiJGJWmOr9pD:-KZYw2GbsiJGJWmOr9pE:b-4fcdb6) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L301)):*
> ```Go
> 	if *geoJSON != "" {
> 		uploadJSON(*geoJSON, batch)
> 		return
> ```

Don't need to return here.

---

*[cmd/dgraphloader/main.go, line 305 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYw3q-qa670o1kXnIa:-KZYw3q-qa670o1kXnIb:b-139e4e) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L305)):*
> ```Go
> 
> 	filesList := strings.Split(*files, ",")
> 	x.AssertTrue(len(filesList) > 0)
> ```

Remove the assert.

---

*[goclient/client/client.go, line 62 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYqK95vRc2yH0wtn0Q:-KZYqK95vRc2yH0wtn0R:b2nreh6) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/client.go#L62)):*
> ```Go
> 		return fmt.Errorf("Atleast one out of objectId and objectValue should be set")
> 	}
> 	if len(nq.ObjectId) == 0 && !hasVal {
> ```

Same as above.

---

*[goclient/client/client.go, line 147 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYqb7RxdEk97JiBVgo:-KZYqb7RxdEk97JiBVgp:b7575hw) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/client.go#L147)):*
> ```Go
> }
> 
> type NQuadOp struct {
> ```

nquadOp struct, so it's internal.

---

*[goclient/client/client.go, line 173 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYqlDaOzwDVA0WMI9t:-KZYqlDaOzwDVA0WMI9u:bz2ukv5) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/client.go#L173)):*
> ```Go
> 	counter := atomic.AddUint64(&batch.mutations, 1)
> RETRY:
> 	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
> ```

Don't need a timeout.

---

*[goclient/client/value.go, line 94 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYrTB85cbwWpDBYEBO:-KZYrTB85cbwWpDBYEBP:b-6nzv6v) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/value.go#L94)):*
> ```Go
> }
> 
> func IsEmpty(val *graph.Value) bool {
> ```

Remove this.

---

*[goclient/client/value.go, line 133 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYsSLoGpQOlHbwu_ku:-KZYsSLoGpQOlHbwu_kv:bgsewgq) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/value.go#L133)):*
> ```Go
> // }
> //
> func ValueFromJson(json string) (Value, error) {
> ```

GeoFromJson(..) or ValueFromGeoJson

---

*[query/graph/graphresponse.proto, line 12 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYt5xROdzxUEKthsjw:-KZYt5xSRFdXa9t_GYEE:b-dhmvnt) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/query/graph/graphresponse.proto#L12)):*
> ```protobuf
> 
> message NQuad {
>     string subject = 1;
> ```

Have a func to convert from uint64 to hex string, so they can fill up these fields.

---

*[rdf/parse.go, line 73 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYtbxEcG8BTTT_rW1G:-KZYtbxEcG8BTTT_rW1H:b6jo31w) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/rdf/parse.go#L73)):*
> ```Go
> 	switch val.Val.(type) {
> 	case *graph.Value_IntVal:
> 		return val.GetIntVal()
> ```

We can just get types.Val struct from this; instead of having 2 functions.

---

*[rdf/parse.go, line 94 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYvSqnOvJ9w9QmYfVN:-KZYvSqnOvJ9w9QmYfVO:b-zi7his) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/rdf/parse.go#L94)):*
> ```Go
> 	// function or the Go client has already set.
> 	t := typeFromValue(nq.ObjectValue)
> 	p := types.ValueForType(t)
> ```

Use this new function mentioned above.

---

*[rdf/parse.go, line 254 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYuTFJYos_HL6iKkL1:-KZYuTFJYos_HL6iKkL2:bhd6zmg) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/rdf/parse.go#L254)):*
> ```Go
> 				// Lets set the object value according to the storage type.
> 				switch t {
> 				case types.StringID:
> ```

Move this switch case to a separate func inside types module; and have both this and the Go client use it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/401)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 9 of 16 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[cmd/dgraphloader/main.go, line 39 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYpvwp8wd-Zdf9BM_J:-KZZ3Vjr8UY20pcr6dhQ:b-hfpv0l) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L39)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

geojson
</blockquote></details>

removed

---

*[cmd/dgraphloader/main.go, line 110 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYwPAQjOPOmowhhQYF:-KZZ3bek4ZrK379RyPor:b-hfpv0l) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L110)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You don't need this.
</blockquote></details>

removed

---

*[cmd/dgraphloader/main.go, line 123 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYwbjjo-3kFRX9_f7E:-KZZ3cqwRlUgzHctgH10:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L123)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Move this func insided the uploadJSON.
</blockquote></details>

Done.

---

*[cmd/dgraphloader/main.go, line 134 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYwi6YUPjg4SI1YOlo:-KZZ3egM849lsmLhZ_z9:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L134)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You don't need this func.
</blockquote></details>

Done.

---

*[cmd/dgraphloader/main.go, line 166 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZ-3w63mydvB22WfbQ:-KZZ3f_YeXauKzuHjGkM:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L166)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Just get rid of all special geo stuff here for now.
</blockquote></details>

Done.

---

*[cmd/dgraphloader/main.go, line 232 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYwxCR-ArPOZXfipdn:-KZZ3gY_1Mv6zpceJM-Z:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L232)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

parseGeoJSON
</blockquote></details>

Done.

---

*[cmd/dgraphloader/main.go, line 291 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYvxcsChfOMZWFN5dn:-KZZ3wjPw91qydJuFxiw:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L291)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

x.Checkf(..)
</blockquote></details>

Done.

---

*[cmd/dgraphloader/main.go, line 296 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYvuhn2UYH_t_Hid_9:-KZZ3wRhdjpqOaPeROMO:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L296)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

The last 2 args were part of flags.
</blockquote></details>

Done.

---

*[cmd/dgraphloader/main.go, line 301 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYw2GbsiJGJWmOr9pD:-KZZ3xXDL0jGmGwhPQAr:b-hfpv0l) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L301)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't need to return here.
</blockquote></details>

removed

---

*[cmd/dgraphloader/main.go, line 305 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYw3q-qa670o1kXnIa:-KZZ3yj-5Ix0mb41ZKdm:b-8bspq3) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/cmd/dgraphloader/main.go#L305)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove the assert.
</blockquote></details>

kept it because we removed geo stuff

---

*[goclient/client/client.go, line 62 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYqK95vRc2yH0wtn0Q:-KZZV42Wa8ccMnv4r9AG:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/client.go#L62)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Done.

---

*[goclient/client/client.go, line 147 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYqb7RxdEk97JiBVgo:-KZZ4NF5L4CNIHgF-SWr:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/client.go#L147)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

nquadOp struct, so it's internal.
</blockquote></details>

Done.

---

*[goclient/client/client.go, line 173 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYqlDaOzwDVA0WMI9t:-KZZ4Q9e1sqCelsgg0g_:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/client.go#L173)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't need a timeout.
</blockquote></details>

Done.

---

*[goclient/client/value.go, line 94 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYrTB85cbwWpDBYEBO:-KZZ4hApdgg9S8W__de3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/value.go#L94)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove this.
</blockquote></details>

Done.

---

*[goclient/client/value.go, line 133 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYsSLoGpQOlHbwu_ku:-KZZ4oHxVPJ_VIhZy94R:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/goclient/client/value.go#L133)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

GeoFromJson(..) or ValueFromGeoJson
</blockquote></details>

Done.

---

*[query/graph/graphresponse.proto, line 12 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYt5xROdzxUEKthsjw:-KZZ5GHyt6ursmaIXDS5:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/query/graph/graphresponse.proto#L12)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Have a func to convert from uint64 to hex string, so they can fill up these fields.
</blockquote></details>

Done.

---

*[rdf/parse.go, line 73 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYtbxEcG8BTTT_rW1G:-KZZUrK5FKZWu6VLVvka:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/rdf/parse.go#L73)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

We can just get types.Val struct from this; instead of having 2 functions.
</blockquote></details>

Done.

---

*[rdf/parse.go, line 94 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYvSqnOvJ9w9QmYfVN:-KZZ6LrTUdxEM2_zKAtF:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/rdf/parse.go#L94)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Use this new function mentioned above.
</blockquote></details>

Done.

---

*[rdf/parse.go, line 254 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYuTFJYos_HL6iKkL1:-KZZUsZ8_fah_vKSGvI9:bxq0pdr) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/rdf/parse.go#L254)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Move this switch case to a separate func inside types module; and have both this and the Go client use it.
</blockquote></details>

Moved it to a function, though go client can't use the function directly as otherwise the user has to set typeID and stuff.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/401)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[cmd/dgraphloader/main.go, line 113 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZhKd2hE4zTWSW_p_f:-KZZhKd2hE4zTWSW_p_g:b-osh4d7) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/cmd/dgraphloader/main.go#L113)):*
> ```Go
> 	c := batch.Counter()
> 	// Lets print an empty line, otherwise Number of Mutations overwrites the previous
> 	// printed line.
> ```

You need a way to stop the printCounters. Otherwise, they can print here and mess up your formatting.

---

*[goclient/client/client.go, line 154 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZiexP5Z3IEcngxkBB:-KZZiexQ2yEzIelTJgem:bqs4slc) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/goclient/client/client.go#L154)):*
> ```Go
> 
> 	nquads   chan nquadOp
> 	requests []*Req
> ```

You don't need to store this.

---

*[goclient/client/client.go, line 180 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZii5Eag80uSO1V-Gs:-KZZii5Eag80uSO1V-Gt:bdvujjj) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/goclient/client/client.go#L180)):*
> ```Go
> 
> func (batch *BatchMutation) makeRequests(req *Req) {
> 	for n := range batch.nquads {
> ```

You can create a req variable here (before the loop starts).

---

*[query/graph/graphresponse.proto, line 12 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYt5xROdzxUEKthsjw:-KZZjL93SpSEbKQ7yVtY:b3ejeo1) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/query/graph/graphresponse.proto#L12)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I don't see this func.

---

*[rdf/parse.go, line 254 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYuTFJYos_HL6iKkL1:-KZZkPKDnN4OHUcvhmh1:bppt92s) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/rdf/parse.go#L254)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Moved it to a function, though go client can't use the function directly as otherwise the user has to set typeID and stuff.
</blockquote></details>

You can pass in both the id and the value as interface{} directly in the function; that way it would be clear to the users as well.

---

*[rdf/parse.go, line 50 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZjPjjD4W1WvZnbrhS:-KZZjPjjD4W1WvZnbrhT:b-7ru42z) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/rdf/parse.go#L50)):*
> ```Go
> }
> 
> func typeValFromValue(val *graph.Value) types.Val {
> ```

typesValFrom(...) // no need for FromValue.

---

*[rdf/parse.go, line 64 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZjab9Um-HUKVtyIYQ:-KZZjab9Um-HUKVtyIYR:byhhce5) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/rdf/parse.go#L64)):*
> ```Go
> 	case *graph.Value_DateVal:
> 		return types.Val{types.DateID, val.GetDateVal()}
> 	case *graph.Value_DatetimeVal:
> ```

Are you missing bytesVal?

---

*[types/conversion.go, line 497 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZk9dKYG4A82E8Y0l5:-KZZk9dKYG4A82E8Y0l6:bcfrjwy) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/types/conversion.go#L497)):*
> ```Go
> }
> 
> func ObjectValue(from Val) (*graph.Value, error) {
> ```

You can take the two arguments: Tid and interface{}. That way, it can be used by the client directly as well. They can just pass in the right type. And then they won't need the helper functions below.

---

*[types/conversion.go, line 530 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZjtuU60XgeG_FME-L:-KZZjtuU60XgeG_FME-M:b-td6z7x) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/types/conversion.go#L530)):*
> ```Go
> 		}
> 		return Datetime(p1.Value.([]byte)), nil
> 	}
> ```

missing bytes again?

---

*[types/value.go, line 9 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZkJhqZ9F9dUIcYEuU:-KZZkJhqZ9F9dUIcYEuV:bsdwl58) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/types/value.go#L9)):*
> ```Go
> 
> // Int returns an int graph.Value
> func Int(val int32) Value {
> ```

You don't need these functions, if you do what I mentioned above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/401)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 9 of 15 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[cmd/dgraphloader/main.go, line 113 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZhKd2hE4zTWSW_p_f:-KZZmbXBrPwbUjH6v-vv:b-pho2ey) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/cmd/dgraphloader/main.go#L113)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You need a way to stop the printCounters. Otherwise, they can print here and mess up your formatting.
</blockquote></details>

Stopping it after flush now.

---

*[goclient/client/client.go, line 154 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZiexP5Z3IEcngxkBB:-KZZmjjCc0GRvL8ni25_:b6mj16l) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/goclient/client/client.go#L154)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You don't need to store this.
</blockquote></details>

that's right, thanks.

---

*[goclient/client/client.go, line 180 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZii5Eag80uSO1V-Gs:-KZZmv65yY0Mp_PIiUcb:b2nr6ik) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/goclient/client/client.go#L180)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You can create a req variable here (before the loop starts).
</blockquote></details>

right that's cleaner.

---

*[query/graph/graphresponse.proto, line 12 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYt5xROdzxUEKthsjw:-KZZn03QEjkCAf0IivBF:b-g6dpmd) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/query/graph/graphresponse.proto#L12)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

I don't see this func.
</blockquote></details>

Its in `goclient/client/value.go` and is called `Uid`

---

*[rdf/parse.go, line 254 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZYuTFJYos_HL6iKkL1:-KZZnkr0Prn_-pIbMybi:b-o2uja5) ([raw file](https://github.com/dgraph-io/dgraph/blob/060d733d823b70729338534787d4d97a1d48e42b/rdf/parse.go#L254)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You can pass in both the id and the value as interface{} directly in the function; that way it would be clear to the users as well.
</blockquote></details>

I am doing that now but I feel `interface{}` isn't very clean.

---

*[rdf/parse.go, line 50 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZjPjjD4W1WvZnbrhS:-KZZnF-YfqGL8JVr2nQe:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/rdf/parse.go#L50)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

typesValFrom(...) // no need for FromValue.
</blockquote></details>

Done.

---

*[rdf/parse.go, line 64 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZjab9Um-HUKVtyIYQ:-KZZnNT4uMuj2PHj2df3:b-s1bt18) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/rdf/parse.go#L64)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Are you missing bytesVal?
</blockquote></details>

Yes, added.

---

*[types/conversion.go, line 497 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZk9dKYG4A82E8Y0l5:-KZZvq6i_mWwTeLC3TXa:bb6un1j) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/types/conversion.go#L497)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You can take the two arguments: Tid and interface{}. That way, it can be used by the client directly as well. They can just pass in the right type. And then they won't need the helper functions below.
</blockquote></details>

Done

---

*[types/conversion.go, line 530 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZjtuU60XgeG_FME-L:-KZZvobov1GTaNAr5nlK:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/types/conversion.go#L530)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

missing bytes again?
</blockquote></details>

Done.

---

*[types/value.go, line 9 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-KZZkJhqZ9F9dUIcYEuU:-KZZtSuSx7Koe35Of6dO:ba3dfob) ([raw file](https://github.com/dgraph-io/dgraph/blob/7f2c76bd6416cff5b0b4aaf7ec60e48e9965c256/types/value.go#L9)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You don't need these functions, if you do what I mentioned above.
</blockquote></details>

Removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/401)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/401#-:-KZ_-xj6stFNWwXOdbSL:bnfp4nl)*
<!-- Sent from Reviewable.io -->
